flagscale/runner/backend.py

```diff
@@ -0,0 +1,2253 @@
+            f.flush()
+            os.fsync(f.fileno())
+        os.chmod(host_stop_script_file, 0o755)
+
+        return host_stop_script_file
+
+
+class TorchBackend(BackendBase):
+    def generate_run_script(self, *args, **kwargs):
+        pass
+
+    def generate_stop_script(self, *args, **kwargs):
+        pass
+
+
+class VllmBackend(BackendBase):
+    def __init__(self, config: DictConfig):
+        super().__init__(config)
+        self.task_type = getattr(self.config.experiment.task, "type", None)
+        assert (
+            self.task_type == "inference" or self.task_type == "serve"
+        ), f"Unsupported task type: {self.task_type}"
+        self._prepare()
+
+    def _prepare(self):
+        if self.task_type == "inference":
+            _update_config_inference(self.config)
+            self.user_args = _get_args_vllm(self.config)
+            self.user_envs = self.config.experiment.get("envs", {})
+            self.user_script = self.config.experiment.task.entrypoint
+            self.resources = parse_hostfile(self.config.experiment.runner.get("hostfile", None))
+        elif self.task_type == "serve":
+            _update_config_serve(self.config)
+            self.user_envs = self.config.experiment.get("envs", {})
+            self.user_args = _get_args_sglang(self.config)
+
+            hostfile_path = self.config.experiment.runner.get("hostfile", None)
+            self.resources = None
+            if hostfile_path:
+                if not os.path.isabs(hostfile_path):
+                    hostfile_path = os.path.join(os.getcwd(), hostfile_path)
+                if os.path.exists(hostfile_path):
+                    self.resources = parse_hostfile(hostfile_path)
+                    for key, value in self.resources.items():
+                        if not value.get("type", None):
+                            logger.warning(
+                                f"The hostfile key type is not set for host {key}, using gpu by default"
+                            )
+                            self.resources[key]["type"] = "gpu"
+
+                    OmegaConf.set_struct(self.config, False)
+                    self.config["nodes"] = list(self.resources.items())
+                    OmegaConf.set_struct(self.config, True)
+                else:
+                    raise ValueError(f"The hostfile {hostfile_path} does not exist")
+
+            if (
+                self.config.experiment.get("runner", {})
+                .get("deploy", {})
+                .get("prefill_decode_disaggregation", False)
+            ):
+                self.user_script = "flagscale/serve/run_disagg_xpyd_router.py"
+            else:
+                self.user_script = "flagscale/serve/run_inference_engine.py"
+
+        logger.info("\n************** configuration **************")
+        logger.info(f"\n{OmegaConf.to_yaml(self.config)}")
+
+    def generate_run_script(self, config, host, node_rank, cmd, background=True, with_test=False):
```

Comment:
This method is too long for test, review and maintenance ... please consider split it into smaller ones.

Reply:
good idea, may be simplified later